### DISCUSSION TITLE
How to check for incorrect casing of specific words?

### DISCUSSION THREAD

**Discussion by bertramakers** [remarkjs/remark#724] [Q&A] [answered]
I want to check for incorrect spelling of a configurable list of product names in the Markdown documentation of the organization I work for, using https://github.com/remarkjs/remark-lint and preferably `remark-cli`.

For example, given the following list:

- `GitHub`
- `remark`
- `Yarn`

I want every occurrence of the same words with a different casing to trigger a warning or error.

For example: `Github`, `github`, `gitHUB` etc should all trigger a warning or error. Same goes for `Remark`, `reMark`, etc.

What would be the easiest way to achieve this with existing plugins/rules?

**Comment by wooorm** [comment, marked as answer]
For spell checking, combine remark with remark-retext and then retext-spell.

For light checkings, use regexes by [writing your own plugin](https://unifiedjs.com/learn/guide/create-a-plugin/) and using [`mdast-util-find-and-replace`](https://github.com/syntax-tree/mdast-util-find-and-replace)

**Reply by wooorm** [reply]
the whole retext ecosystem has tons of existing tools btw, for checking stuff!

**Reply by bertramakers** [reply]
Thanks for the info! I have very little (no) experience with writing my own plugins for retext so I will need to plan some time to look into it 😄

**Reply by bertramakers** [reply]
Hi @wooorm . I was revisiting this because I feel a bit more familiar with unified plugins now, especially after our conversation in https://github.com/remarkjs/remark-gfm/issues/42 and fixing that issue myself in a PR. The article on how to write my own plugin to check for spelling mistakes of known words seems pretty straightforward now. But what I don't find anywhere is how I could make my plugin also fix incorrect spellings when the `-o` option is used in `remark-cli`. Do you also have an article on that or some tips to get started on that?

**Reply by wooorm** [reply]
Changes to the AST that retext uses, when in remark with `remark-retext`, can’t be changed back into the AST that remark itself uses.
It’s a very complex problem, some markdown might contain `*foo*sball and such`, but what to do if the word was replaced to `tennis`? What if `foos` was removed? What if something was added at the start?

There’s al alternate way: only parse the document, record changes (as vfile-messages, particular with an `expected` value), but *don’t* output the AST. Then, from end-to-start apply those replacements using the positional info on those messages. This is what ESLint also does. It’s fine for small things (adding a semicolon), but gets complex for multiple replacements and for longer changes. And, it can’t be combined with AST-based changes.
